# Pandas fundamentals

#### Written for the QuantEcon Indian Summer Workshop (August 2022)
#### Author: [Shu Hu](https://shu-hu.com/intro.html)

In [1]:
import pandas as pd

Given the dataframe below you are required to do the following exercises.

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/QuantEcon/lecture-python-programming/master/source/_static/lecture_specific/pandas/data/test_pwt.csv')

In [3]:
df

country country isocode  year          POP       XRAT         tcgdp  \
0      Argentina             ARG  2000    37335.653   0.999500  2.950722e+05   
1      Australia             AUS  2000    19053.186   1.724830  5.418047e+05   
2          India             IND  2000  1006300.297  44.941600  1.728144e+06   
3         Israel             ISR  2000     6114.570   4.077330  1.292539e+05   
4         Malawi             MWI  2000    11801.505  59.543808  5.026222e+03   
5   South Africa             ZAF  2000    45064.098   6.939830  2.272424e+05   
6  United States             USA  2000   282171.957   1.000000  9.898700e+06   
7        Uruguay             URY  2000     3219.793  12.099592  2.525596e+04   

          cc         cg  
0  75.716805   5.578804  
1  67.759026   6.720098  
2  64.575551  14.072206  
3  64.436451  10.266688  
4  74.707624  11.658954  
5  72.718710   5.726546  
6  72.347054   6.032454  
7  78.978740   5.108068

### Exercise 1 (View data)

### Exercise 1.1

Show the top 5 rows of the dataframe ``df``.

### Solution

In [4]:
df.head(5)

country country isocode  year          POP       XRAT         tcgdp  \
0  Argentina             ARG  2000    37335.653   0.999500  2.950722e+05   
1  Australia             AUS  2000    19053.186   1.724830  5.418047e+05   
2      India             IND  2000  1006300.297  44.941600  1.728144e+06   
3     Israel             ISR  2000     6114.570   4.077330  1.292539e+05   
4     Malawi             MWI  2000    11801.505  59.543808  5.026222e+03   

          cc         cg  
0  75.716805   5.578804  
1  67.759026   6.720098  
2  64.575551  14.072206  
3  64.436451  10.266688  
4  74.707624  11.658954

### Exercise 1.2

Show the bottom 5 rows of the dataframe ``df``.

### Solution

In [5]:
df.tail(5)

country country isocode  year         POP       XRAT         tcgdp  \
3         Israel             ISR  2000    6114.570   4.077330  1.292539e+05   
4         Malawi             MWI  2000   11801.505  59.543808  5.026222e+03   
5   South Africa             ZAF  2000   45064.098   6.939830  2.272424e+05   
6  United States             USA  2000  282171.957   1.000000  9.898700e+06   
7        Uruguay             URY  2000    3219.793  12.099592  2.525596e+04   

          cc         cg  
3  64.436451  10.266688  
4  74.707624  11.658954  
5  72.718710   5.726546  
6  72.347054   6.032454  
7  78.978740   5.108068

### Exercise 1.3

Show a quick statistic summary of the dataframe ``df``.

### Solution

In [6]:
df.describe()

year           POP       XRAT         tcgdp         cc         cg
count     8.0  8.000000e+00   8.000000  8.000000e+00   8.000000   8.000000
mean   2000.0  1.763826e+05  16.415811  1.606312e+06  71.404995   8.145477
std       0.0  3.479223e+05  22.758175  3.397025e+06   5.318015   3.383397
min    2000.0  3.219793e+03   0.999500  5.026222e+03  64.436451   5.108068
25%    2000.0  1.037977e+04   1.543623  1.032544e+05  66.963157   5.689611
50%    2000.0  2.819442e+04   5.508580  2.611573e+05  72.532882   6.376276
75%    2000.0  1.043411e+05  20.310094  8.383896e+05  74.959919  10.614755
max    2000.0  1.006300e+06  59.543808  9.898700e+06  78.978740  14.072206

### Exercise 1.4

Transpose the dataframe ``df``.

### Solution

In [7]:
df.T

0            1             2             3  \
country             Argentina    Australia         India        Israel   
country isocode           ARG          AUS           IND           ISR   
year                     2000         2000          2000          2000   
POP                 37335.653    19053.186   1006300.297       6114.57   
XRAT                   0.9995      1.72483       44.9416       4.07733   
tcgdp            295072.21869  541804.6521  1728144.3748  129253.89423   
cc                  75.716805    67.759026     64.575551     64.436451   
cg                   5.578804     6.720098     14.072206     10.266688   

                           4             5              6             7  
country               Malawi  South Africa  United States       Uruguay  
country isocode          MWI           ZAF            USA           URY  
year                    2000          2000           2000          2000  
POP                11801.505     45064.098     282171.957      3219.793  
XRAT               59.543808       6.93983            1.0     12.099592  
tcgdp            5026.221784  227242.36949      9898700.0  25255.961693  
cc                 74.707624      72.71871      72.347054      78.97874  
cg                 11.658954      5.726546       6.032454      5.108068

### Exercise 1.5

Sort the dataframe ``df`` by values in column ``tcgdp`` ascendingly.

### Solution

In [8]:
df.sort_values(by="tcgdp")

country country isocode  year          POP       XRAT         tcgdp  \
4         Malawi             MWI  2000    11801.505  59.543808  5.026222e+03   
7        Uruguay             URY  2000     3219.793  12.099592  2.525596e+04   
3         Israel             ISR  2000     6114.570   4.077330  1.292539e+05   
5   South Africa             ZAF  2000    45064.098   6.939830  2.272424e+05   
0      Argentina             ARG  2000    37335.653   0.999500  2.950722e+05   
1      Australia             AUS  2000    19053.186   1.724830  5.418047e+05   
2          India             IND  2000  1006300.297  44.941600  1.728144e+06   
6  United States             USA  2000   282171.957   1.000000  9.898700e+06   

          cc         cg  
4  74.707624  11.658954  
7  78.978740   5.108068  
3  64.436451  10.266688  
5  72.718710   5.726546  
0  75.716805   5.578804  
1  67.759026   6.720098  
2  64.575551  14.072206  
6  72.347054   6.032454

### Exercise 2 (Select data)

### Exercise 2.1

Select the column ``tcgdp`` from the dataframe ``df``, yielding a series.

### Solution

In [9]:
df["tcgdp"]

0    2.950722e+05
1    5.418047e+05
2    1.728144e+06
3    1.292539e+05
4    5.026222e+03
5    2.272424e+05
6    9.898700e+06
7    2.525596e+04
Name: tcgdp, dtype: float64

In [10]:
df.tcgdp

0    2.950722e+05
1    5.418047e+05
2    1.728144e+06
3    1.292539e+05
4    5.026222e+03
5    2.272424e+05
6    9.898700e+06
7    2.525596e+04
Name: tcgdp, dtype: float64

### Exercise 2.2

Select the top 3 rows from the dataframe ``df``.

### Solution

In [11]:
df[0:3]

country country isocode  year          POP      XRAT         tcgdp  \
0  Argentina             ARG  2000    37335.653   0.99950  2.950722e+05   
1  Australia             AUS  2000    19053.186   1.72483  5.418047e+05   
2      India             IND  2000  1006300.297  44.94160  1.728144e+06   

          cc         cg  
0  75.716805   5.578804  
1  67.759026   6.720098  
2  64.575551  14.072206

### Exercise 2.3


Get the first row of data across all the columns.

### Solution

In [12]:
df.loc[0]

country               Argentina
country isocode             ARG
year                       2000
POP                   37335.653
XRAT                     0.9995
tcgdp              295072.21869
cc                    75.716805
cg                     5.578804
Name: 0, dtype: object

### Exercise 2.4

Select all values on two columns ``country`` and ``POP``.

### Solution

In [13]:
df[["country", "POP"]]

country          POP
0      Argentina    37335.653
1      Australia    19053.186
2          India  1006300.297
3         Israel     6114.570
4         Malawi    11801.505
5   South Africa    45064.098
6  United States   282171.957
7        Uruguay     3219.793

You can also use the ``loc`` method which gives you access to locators for both column and row selection.

In [14]:
df.loc[:, ["country", "POP"]]

country          POP
0      Argentina    37335.653
1      Australia    19053.186
2          India  1006300.297
3         Israel     6114.570
4         Malawi    11801.505
5   South Africa    45064.098
6  United States   282171.957
7        Uruguay     3219.793

### Exercise 2.5

Select values in rows 2-4 on two columns ``country`` and ``POP``.

### Solution

In [15]:
df.loc[2:4, ["country", "POP"]]

country          POP
2   India  1006300.297
3  Israel     6114.570
4  Malawi    11801.505

### Exercise 2.6

Get a scalar value for the 2nd row in the column ``POP`` which should be ``1006300.297``.

### Solution

In [16]:
df.loc[2, "POP"]

1006300.297

In [17]:
# The .at property acts only on single value pairs for columns and rows, 
# while .loc can include ranges and more advanced filters
df.at[2, "POP"]

1006300.297

### Exercise 3 (Boolean indexing)

### Exercise 3.1 

Select data with column ``POP``'s values greater than ``30_000``.

### Solution

In [18]:
df[df['POP'] > 30_000]

country country isocode  year          POP      XRAT         tcgdp  \
0      Argentina             ARG  2000    37335.653   0.99950  2.950722e+05   
2          India             IND  2000  1006300.297  44.94160  1.728144e+06   
5   South Africa             ZAF  2000    45064.098   6.93983  2.272424e+05   
6  United States             USA  2000   282171.957   1.00000  9.898700e+06   

          cc         cg  
0  75.716805   5.578804  
2  64.575551  14.072206  
5  72.718710   5.726546  
6  72.347054   6.032454

### Exercise 3.2

We see

In [19]:
df1 = df.copy()
df1

country country isocode  year          POP       XRAT         tcgdp  \
0      Argentina             ARG  2000    37335.653   0.999500  2.950722e+05   
1      Australia             AUS  2000    19053.186   1.724830  5.418047e+05   
2          India             IND  2000  1006300.297  44.941600  1.728144e+06   
3         Israel             ISR  2000     6114.570   4.077330  1.292539e+05   
4         Malawi             MWI  2000    11801.505  59.543808  5.026222e+03   
5   South Africa             ZAF  2000    45064.098   6.939830  2.272424e+05   
6  United States             USA  2000   282171.957   1.000000  9.898700e+06   
7        Uruguay             URY  2000     3219.793  12.099592  2.525596e+04   

          cc         cg  
0  75.716805   5.578804  
1  67.759026   6.720098  
2  64.575551  14.072206  
3  64.436451  10.266688  
4  74.707624  11.658954  
5  72.718710   5.726546  
6  72.347054   6.032454  
7  78.978740   5.108068

In [20]:
df1["X"] = ["A", "B", "C", "D", "E", "F", "G","H"]

In [21]:
df1

country country isocode  year          POP       XRAT         tcgdp  \
0      Argentina             ARG  2000    37335.653   0.999500  2.950722e+05   
1      Australia             AUS  2000    19053.186   1.724830  5.418047e+05   
2          India             IND  2000  1006300.297  44.941600  1.728144e+06   
3         Israel             ISR  2000     6114.570   4.077330  1.292539e+05   
4         Malawi             MWI  2000    11801.505  59.543808  5.026222e+03   
5   South Africa             ZAF  2000    45064.098   6.939830  2.272424e+05   
6  United States             USA  2000   282171.957   1.000000  9.898700e+06   
7        Uruguay             URY  2000     3219.793  12.099592  2.525596e+04   

          cc         cg  X  
0  75.716805   5.578804  A  
1  67.759026   6.720098  B  
2  64.575551  14.072206  C  
3  64.436451  10.266688  D  
4  74.707624  11.658954  E  
5  72.718710   5.726546  F  
6  72.347054   6.032454  G  
7  78.978740   5.108068  H

Filter rows containing ``A``, ``C``, ``G`` in column ``X`` using method [isin()](https://pandas.pydata.org/docs/reference/api/pandas.Series.isin.html#pandas.Series.isin).

### Solution

In [22]:
df1[df1["X"].isin(["A", "C", "G"])]

country country isocode  year          POP     XRAT         tcgdp  \
0      Argentina             ARG  2000    37335.653   0.9995  2.950722e+05   
2          India             IND  2000  1006300.297  44.9416  1.728144e+06   
6  United States             USA  2000   282171.957   1.0000  9.898700e+06   

          cc         cg  X  
0  75.716805   5.578804  A  
2  64.575551  14.072206  C  
6  72.347054   6.032454  G